packages/components/overlay/src/overlay.js

```diff
@@ -0,0 +1,197 @@
+			const { verticalPosition, maxHeight } = this.calculateVerticalPosition(anchorRect, contentHeight);
+
+			Object.assign(container.style, horizontalPosition, verticalPosition);
+			Object.assign(contentContainer.style, { maxHeight });
+		});
+	}
+
+	/**
+	 * Calculates horizontal position and width of the container
+	 * @private
+	 * @param {!DomRect} anchorRect ClientRect of the container's anchor
+	 * @param {!Number} contentWidth width of the content
+	 * @returns {!Object} position on left, right and width of the container
+	 */
+	calculateHorizontalPosition(anchorRect, contentWidth) {
+		const result = { left: 'auto', right: 'auto', width: undefined };
+		const { width, left, right } = anchorRect;
+		if (this.autoWidth) {
+			// calculate horizontal position
+			const menuWidth = anchorRect.width;
+			const { innerWidth } = window;
+			const fitOnRight = left + contentWidth + HALF_UNIT <= innerWidth;
+			const fitOnLeft = left + width - contentWidth - HALF_UNIT >= 0;
+			if (this.dir === 'rtl') {
+				if (!fitOnRight && !fitOnLeft) {
+					// cannot fit on both sides, align to left border
+					result.right = `${innerWidth - HALF_UNIT - contentWidth}px`;
+				} else if (!fitOnLeft) {
+					// cannot fit on left, put on right side
+					result.left = `${left}px`;
+				} else {
+					/// put on left as default
+					result.right = `${innerWidth - right}px`;
+				}
+			} else {
+				if (!fitOnRight && !fitOnLeft) {
+					// cannot fit on both sides, align to right border
+					result.left = `${innerWidth - contentWidth - HALF_UNIT - left}px`;
+				} else if (!fitOnRight) {
+					// cannot fit on right, put on left side
+					result.left = `${left + menuWidth - contentWidth}px`;
+				} else {
+					// put on right as default
+					result.left = left + 'px';
+				}
+			}
+		} else {
+			result.left = left + 'px';
+			result.width = width + 'px';
+		}
+		return result;
+	}
+
+	/**
+	 * Calculates vertical position and max height of the container
+	 * @private
+	 * @param {!DomRect} anchorRect
+	 * @param {!Number} contentHeight
+	 * @returns {!Object} position on top, bottom and max height of the container
+	 */
+	calculateVerticalPosition(anchorRect, contentHeight) {
+		const verticalPosition = { top: 'auto', bottom: 'auto' };
+		const { top, bottom, height } = anchorRect;
+		const { innerHeight } = window;
+		let maxHeight = 0;
+		const paddingTop = top - HALF_UNIT;
+		const paddingBottom = innerHeight - bottom - HALF_UNIT;
+		if (paddingBottom < contentHeight && paddingBottom < paddingTop) {
+			// When we don't have enough space below and have more space above, we put dropdown above.
+			maxHeight = top > MAX_HEIGHT ? MAX_HEIGHT : paddingTop;
+			if (maxHeight < contentHeight) {
+				// maxHeight will be set as a new height of a content container after position update
+				contentHeight = maxHeight;
+			}
+			verticalPosition.top = top - height - contentHeight + HALF_UNIT + 'px';
```

Comment:
We were talking about that this is weird that it works (or then line 163), it seems like this should not work with the height being included in both top and bottom calculations.

Reply:
I found the problem:
the `top` value actually equals the length between top window border and top border of the anchor: https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect

So, the calculation for the bottom positioning (line 163) is correct. But the calculation for the top positioning was wrong and looks fine just because height was near 20px, and `-height + HALF_UNIT` was almost equal to `-HALF_UNIT`. So the correct one should be: (top) - (height of the content) - (gap between anchor and dropdown == HALF_UNIT).